docs/glossary.md

```diff
@@ -0,0 +1,101 @@
+3. Sleight of hand!  A naive evaluator was too slow (at least decades ago), so behind the scenes a lot is optimized & cached.  When not stepping, most of the execution _does_ happen in hidden state, but care was taken to make this equivalent.
+
+## explicit control evaluator
+
+The `boxer-eval` state machine maintains its state in explicit global variables (one can think of them as registers) and stacks.
+Procedure calls in boxer grow boxer's stack, not the underlying Lisp's stack.
+
+This exposes the state for stepper / debugging / error handling.  It also allows some forms of context switching between "processes".
+
+### pdl (from archaic "Push Down List")
+
+This is the main control stack.
+
+It consists of typed frames, each saving/restoring a different subset of state.
+
+### vpdl (from "Value Push Down List")
+
+A stack of values which will become arguments to functions.
+
+The naive "copy and execute" model actually mutates the calling box, replacing calls with their return values, variables with their values etc.
+As an optimization when not stepping, this can mostly be skipped — and VPDL stack holds the intermediate values.
+
+It's also the interface to primitives.  `defboxer-primitive` looks similar to `defun` but constructs a lisp functions that pops arguments from VPDL, and pushes return value to VPDL.
+
+### recursive evaluation
+
+The evaluator calls primitives.  Primitives like `repeat` or `tell` that need to run boxer code can't just call the evaluator (that would grow the lisp stack and would be opaque to the stepper).
+
+`defrecursive-eval-primitive` / `defrecursive-funcall-primitive` macros split each such primitive into a before, after, and cleanup parts, plus a custom PDL frame type.  Between the parts, control returns to the main eval loop, with state (e.g. `repeat`'s loop counter) retained in the PDL frame.
+
+### process
+
+### sfun
+
+### ufun
+
+## vc - virtual copy
+
+## Variables in Boxer
+
+All this applies equally to variable and procedure lookup by name, the rules are the same.
+(Lookup finds a named box; only then you'll see if it's a data or doit box, and what to do with it.
+In this sense, Boxer falls into ["Lisp-1" camp](http://www.nhplace.com/kent/Papers/Technical-Issues.html).)
+
+### Static
+
+Under the "copy and execute" model, a called box is copied into the calling box.
+Names are looked up directly in the hierarchy of boxes containing the current box.
+This naturally resulting semantic is dynamic binding — free variables are can be bound by the caller.
+
+That's what "static" lookup refers to — search up the hierarchy of materialized boxes.
+
+(The term has nothing to do with C/Java meaning of "static" lifetime.)
+
+### Dynamic
+
+As an optimization, when not stepping, the called box may not be copied into the call site.
+But it has to resolve variables _as if_ it was there.  So a separate dynamic bindings stack is maintained, and "dynamic" lookup refers to checking that stack.
```

Comment:
Here I have less confidence, I skimmed a lot of the variables code...  Is it accurate to describe it as a stack?

Reply:
I would call it "a" stack, of some kind, even if it is a materialised incarnation of that typically hidden construct. In Infusion, a semi-materialised system, I did call such a slice as a stack as in "getThatStack" - https://github.com/fluid-project/infusion/blob/main/src/framework/core/js/FluidIoC.js#L1135 . Concept-stretching perhaps but these signposts can be useful in showing us where we have been and where we are going.

Reply:
yep, Klotz1989 p32 sub-section "Dynamic Variables Stack" :-)

The description there is of an alist of alists which is costly to GC; p33 says "A future implementation might use two simple vectors for this association list, and eliminate temporary storage allocation except in deep recursion." — and in src/evaluator/bind.lisp I see that's been implemented (good comments there).  Anyway, conceptually it is a stack.